x-pack/metricbeat/module/openai/usage/usage.go

```diff
@@ -0,0 +1,392 @@
+	if err != nil {
+		return nil, fmt.Errorf("creating state store: %w", err)
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		logger:        logp.NewLogger("openai.usage"),
+		config:        config,
+		stateStore:    st,
+	}, nil
+}
+
+// Fetch method implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+//
+//  1. Creates a rate-limited HTTP client with configured timeout and burst settings
+//  2. Sets up the time range for data collection:
+//     i.	End date is current UTC time for realtime collection
+//     ii.	End date is previous day for non-realtime collection
+//     iii. Start date is calculated based on configured lookback days
+//  3. Fetches usage data for each day in the date range
+func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+	httpClient := newClient(
+		context.TODO(),
+		m.logger,
+		rate.NewLimiter(
+			rate.Every(time.Duration(*m.config.RateLimit.Limit)*time.Second),
+			*m.config.RateLimit.Burst,
+		),
+		m.config.Timeout,
+	)
+
+	m.report = report
+
+	endDate := time.Now().UTC()
+
+	if !m.config.Collection.Realtime {
+		// If we're not collecting realtime data, then just pull until
+		// yesterday (in UTC).
+		endDate = endDate.AddDate(0, 0, -1)
+	}
+
+	startDate := endDate.AddDate(0, 0, -m.config.Collection.LookbackDays)
+
+	return m.fetchDateRange(startDate, endDate, httpClient)
+}
+
+// fetchDateRange retrieves OpenAI API usage data for each configured API key within a specified date range.
+//
+// For each API key:
+// 1. Generates a secure SHA-256 hash of the key for state tracking
+// 2. Checks the state store for the last processed date
+// 3. Adjusts start date if previous state exists to avoid duplicate collection
+// 4. Iterates through each day in the range, collecting usage data
+// 5. Updates state store with the latest processed date
+func (m *MetricSet) fetchDateRange(startDate, endDate time.Time, httpClient *RLHTTPClient) error {
+	for _, apiKey := range m.config.APIKeys {
+		// SHA-256 is a cryptographic hash function that generates a 256-bit (32-byte) digest,
+		// represented as a 64-character hexadecimal string. The hash function is deterministic,
+		// ensuring the same input consistently produces identical output, while being computationally
+		// infeasible to reverse. Its strong collision resistance makes it highly suitable for
+		// secure key storage.
+		//
+		// The hexadecimal representation uses only alphanumeric characters [0-9a-f], making it
+		// ideal for cross-platform filename compatibility. The fixed-length output provides
+		// predictable storage requirements and consistent behavior across different systems.
+		hasher := sha256.New()
+		hasher.Write([]byte(apiKey.Key))
+		hashedKey := hex.EncodeToString(hasher.Sum(nil))
+		stateKey := "state_" + hashedKey
```

Comment:
it could also be done at the module initialisation, avoiding recalculating it every time

Reply:
changed the logic with lookup + hashing.